### PR TITLE
add current layer to route options

### PR DESCRIPTION
**Pull request by Zayankovsky**
### Changelog

```
<changelog>Added current layer to route options for reroute and alternatives requests</changelog>
```

### CODE REVIEW

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/factory/RouteProgressFactory.kt [hunk 1]
```diff
@@ -55,7 +56,8 @@ object RouteProgressFactory {
         fractionTraveled: Float,
         remainingWaypoints: Int,
         upcomingRoadObjects: List<UpcomingRoadObject>,
-        stale: Boolean
+        stale: Boolean,
+        layer: Int?,
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/factory/RouteProgressFactory.kt:60]: Could we call the value itself something like `zLevel`? `Layer` is a very overloaded term in a Mapbox ecosystem.

Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/factory/RouteProgressFactory.kt:60]: I think we should also expose a getter for this value through `MapboxNavigation`.

Typically when a route request is made, it's not done from the `RouteProgress` callback, it's trigger by some user interaction, so developers will come looking for this value and will have to cache it themselves anyway.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteProgress.kt [hunk 2]
```diff
@@ -36,6 +36,7 @@ import com.mapbox.navigation.base.trip.model.roadobject.UpcomingRoadObject
  * @param upcomingRoadObjects list of upcoming road objects.
  * @param stale `true` if there were no location updates for a significant amount which causes
  * a lack of confidence in the progress updates being sent.
+ * @param zLevel [Int] current Z-level. Can be used to to build route from proper level of road.
```
Comment (korshaknn) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteProgress.kt:39]:
```suggestion
 * @param zLevel [Int] current Z-level. Can be used to build a route from a proper level of a road.
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/factory/RouteProgressFactory.kt [hunk 3]
```diff
@@ -40,6 +40,7 @@ object RouteProgressFactory {
      * @param upcomingRoadObjects list of upcoming road objects.
      * @param stale `true` if there were no location updates for a significant amount which causes
      * a lack of confidence in the progress updates being sent.
+     * @param zLevel [Int] current Z-level. Can be used to to build route from proper level of road.
```
Comment (korshaknn) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/factory/RouteProgressFactory.kt:43]:
```suggestion
     * @param zLevel [Int] current Z-level. Can be used to build a route from a proper level of a road.
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteProgress.kt [hunk 4]
```diff
@@ -120,7 +124,8 @@ class RouteProgress internal constructor(
             "fractionTraveled=$fractionTraveled, " +
             "remainingWaypoints=$remainingWaypoints, " +
             "upcomingRoadObjects=$upcomingRoadObjects, " +
-            "stale=$stale" +
+            "stale=$stale," +
```
Comment (korshaknn) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteProgress.kt:127]:
```suggestion
            "stale=$stale, " +
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 5]
```diff
@@ -502,6 +502,13 @@ class MapboxNavigation(
      */
     fun getNavigationSessionState(): NavigationSessionState = navigationSession.state
 
+    /**
+     * Provides the current Z-Level. Can be used to to build route from proper level of road.
```
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:506]:
```suggestion
     * Provides the current Z-Level. Can be used to build a route from a proper level of a road.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/MapboxRouteOptionsUpdater.kt [hunk 6]
```diff
@@ -127,6 +127,16 @@ class MapboxRouteOptionsUpdater : RouteOptionsUpdater {
...
                         )
                     )
+                    .layersList(
+                        mutableListOf(routeProgress.zLevel).apply {
+                            val legacyLayerList = routeOptions.layersList()
+                            if (legacyLayerList != null) {
+                                addAll(legacyLayerList.takeLast(remainingWaypoints))
+                            } else {
+                                repeat(remainingWaypoints) { add(null) }
+                            }
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/MapboxRouteOptionsUpdater.kt:137]: The docs say:
> If provided, the list of layers must be the same length as the list of coordinates.

With the proposed impl, the provided list would be of size `coordinates - 1`. We need to fetch the current `zLevel` from the `RouteProgress`/`TripSession` and add it as the first element in this list.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/MapboxRouteOptionsUpdater.kt:137]: Ah you did that above already, never mind!

**Review by LukasPaczos — APPROVED** [review]
LGTM

**Review by korshaknn — APPROVED** [review]

**Pull request merged**